src/python/pants/backend/python/util_rules/vcs_versioning.py

```diff
@@ -99,8 +111,49 @@ async def generate_python_from_setuptools_scm(
     return GeneratedSources(output_snapshot)
 
 
+# This is only used to register our implementation with the plugin hook via unions.
+class PythonVCSVersionMappingMarker(FirstPartyPythonMappingImplMarker):
+    pass
+
+
+class VCSVersionPythonResolveField(PythonResolveField):
+    alias = "python_resolve"
+
+
+@rule
+async def map_to_python_modules(
+    all_targets: AllTargets,
```

Comment:
It's better caching semantics to have the `AllTargets` handling be a dedicated rule. That way, if a Go target changed but didn't resolve in a VCS target changing, there is no need to recompute the module mapping.

See https://github.com/pantsbuild/pants/issues/14720.

Reply:
That would have been useful to have in a comment in the code I cargo-culted from...

Reply:
But it makes total sense!